### ISSUE TITLE
go-torch is incorrect for some profiles

### ISSUE THREAD

**Issue by prashantv** [uber/go-torch#28]
See attached profile + generated torch.

I'm seeing:
![image](https://cloud.githubusercontent.com/assets/140159/9398622/c3f0151e-475d-11e5-81fd-49a7f53d27d1.png)

The RelayFrame (1.82 seconds) is wider than (*defaultFramePool).Get which took 13.64 seconds.

https://drive.google.com/a/uber.com/file/d/0B4HJl0NtpbKNcTZ1TGNXcFp6MTA/view?usp=sharing

**Comment by sandlerben** [comment]
I don't have permission to see the profile.

**Comment by prashantv** [comment]
Shared

**Comment by prashantv** [comment]
I wrote a quick pprof -> stack outputter using the pprof raw output, and then used stackcollapse etc and got the following flame graph:
http://prashantv.com/files/torch.svg

I think that using the raw output serves us better as we get all the samples and can drill down, and there's no chance of losing anything. I'll clean it up and send out a PR when I get some free time.

**Comment by sandlerben** [comment]
Cool. It looks like go tool pprof has quite a bit of logic when it is
reading the raw output. Are we sure we want to handle all of this?

How did you go from the raw output to stack traces?

**Comment by sandlerben** [comment]
Ok, so there is something fishy going on in graph.go. Right now, we generate the folded stacks by doing a DFS traversal of the dot graph output of `go tool pprof`. The folded stacks (which we directly feed to the flame graph script) look like:
```
function1;function2;function3 3
function4;function5 2
```
The current logic must not work in all cases. go-torch traverses every root->leaf path in the graph, and that becomes a line in the folded stacks output. However, consider the following case:

function1 calls function2 5 times
function1 calls function3 2 times
function3 calls function4 once
function3 calls function5 once

When go-torch traverses this graph, the following will be output:
```
function1;function2 5
function1;function3;function4 3
function1;function3;function5 3
```
However, the following *should* be output:
```
function1;function2 5
function1;function3;function4 1
function1;function3;function5 1
```

Here is the problem: when go-torch traverses a path, it maintains the _sum_ weight of that path. So the path function1->function3->function4 has _sum_ weight of 3. In reality, we should only care about the weight of the final edge, right? Here is the code I am referencing: https://github.com/uber/go-torch/blob/master/graph/graph.go#L213

Can you confirm my thinking on what should be output vs what is output?

**Comment by sandlerben** [comment]
Also, if this works, I would much prefer to continue using `go tool pprof` than read the stack traces directly.

**Comment by prashantv** [comment]
@sandlerben Before I look into the details of the DFS, let me explain my approach which I think is simpler easier to validate.

The output of `go tool pprof --raw pprof.pb.gz` is:
```
Samples:
samples/count cpu/nanoseconds
          1   10000000: 1 2 3 4
          1   10000000: 5 6 2 3 4
          1   10000000: 7 8 2 3 4
          1   10000000: 1 2 3 4
          1   10000000: 9 2 3 4
          1   10000000: 10 2 3 4
[...]
Locations
     1: 0x41e5db runtime.scanobject :0 s=0
     2: 0x41e121 runtime.gcDrainUntilPreempt :0 s=0
     3: 0x41ac73 runtime.gcBgMarkWorker :0 s=0
[...]
```

The `Samples` output is: `number of samples | total nanoseconds | function IDs (stack)`

The numbers in the stack are references to functions, the symbol of which is in `Locations`, so you can convert the samples to stack pretty trivially, so I wrote a simpler parser which will output the full stack + count, so that outputs something like:
```
syscall.Syscall
syscall.write
syscall.Write
net.(*netFD).Write
net.(*conn).Write
github.com/uber/tchannel/golang.(*Frame).WriteOut
github.com/uber/tchannel/golang.(*Connection).writeFrames
runtime.goexit
124
[...]
```

I then passed to stack collapse script in FlameGraph and then generate a Flame graph out of it. I can do the stack collapse inside of the Go code since it's pretty trivial, so then we'd just need to pass it to FlameGraph.

The output from `go tool pprof --raw` matches the output that `flamegraph.pl` expects very closely, so I think it's better to use this output directly and keep it simple rather than using all the graph logic.

What do you think?

**Comment by sandlerben** [comment]
That sounds great.

**Comment by sandlerben** [comment]
Can I help out? If you think this can wait a week or two, I'd love to do this fix.

**Comment by prashantv** [comment]
Sure! I pushed a branch `raw` with my functional changes, but it could probably be done cleaner and with tests.
https://github.com/uber/go-torch/compare/raw?expand=1&ts=2

The things I'd like to see are:
 - integrate stackcollapse into Go
 - use streams everywhere (the input could be streamed from pprof and processed rather than as a string)
 - remove interfaces and just test using the streams directly

The branch is functional enough to unblock me so no rush.

**Comment by sandlerben** [comment]
@prashantv this issue was resolved in #29

**Comment by prashantv** [comment]
Yep, fixed in #29.
